src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -116,79 +123,142 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
-                    }
-                });
+                    // If we detect that the client is offline. Simply stop trying to send the payload.
+                    // When the client reconnects we'll resend it.
+                    return;
+                }
+
+                Log.BeginUpdateDisplayAsync(_logger, _client.ConnectionId);
+                await _client.SendAsync("JS.RenderBatch", Id, pending.BatchId, pending.Data);
             }
-            catch (Exception syncException)
+            catch (Exception e)
             {
-                pendingRenderInfo.TrySetException(syncException);
+                Log.SendBatchDataFailed(_logger, e);
             }
 
-            // When the render is completed (success, fail, or timeout), stop tracking it
-            return pendingRenderInfo.Task.ContinueWith(task =>
-            {
-                _pendingRenders.TryRemove(renderId, out var ignored);
-                if (task.IsFaulted)
-                {
-                    UnhandledException?.Invoke(this, task.Exception);
-                }
-            });
+            // We don't have to remove the entry from the list of pending batches if we fail to send it or the client fails to
+            // acknowledge that it received it. We simply keep it in the queue until we receive another ack from the client for
+            // a later batch (clientBatchId > thisBatchId) or the circuit becomes disconnected and we ultimately get evicted and
+            // disposed.
         }
 
-        public void OnRenderCompleted(long renderId, string errorMessageOrNull)
+        public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
         {
-            if (_pendingRenders.TryGetValue(renderId, out var pendingRenderInfo))
+            if (_disposing)
+            {
+                // Disposing so don't do work.
+                return;
+            }
+
+            // We peek first in case we are receiving a duplicate or out of order ack from
+            // the client.
+            while (PendingRenderBatches.TryPeek(out var entry))
             {
-                if (errorMessageOrNull == null)
+                // Dequeue entries until we sync with the render batch number received.
+                if (incomingBatchId < entry.BatchId)
                 {
-                    pendingRenderInfo.TrySetResult(null);
+                    _logger.LogInformation($"Incoming batch {incomingBatchId} already acknowledged.");
+                    // We successfully received an ack for a pending batch left.
+                    // Which we already processed so no work.
+                    break;
                 }
-                else
+
+                // batchId >= entry.BatchId -> We are in sync or we missed an ack from the client.
+                // We simply catch up as a bigger number means the client already rendered
+                // previous batches successfully.
+                if (incomingBatchId >= entry.BatchId)
```

Comment:
This is guaranteed to be true since you check for the converse in line 196

Reply:
removed

Reply:
Shouldn't it be a fatal error to observe a gap in the sequence of ACKs from the client? There's no scenario where we should be able to receive an ACK for, say, batch 15 unless we've also received ACKs for 1-14. If that happens, something bad is going on, so shouldn't we error out so we have a chance of recognizing and finding the bug?

This is based on the fact that, if some ACK got lost due to a disconnection, then on reconnection the process is: server re-sends all unacknowledged batches, client (re)sends acknowledgement for all those batches.

Reply:
In other words, I'd expect this condition to be `if (incomingBatchId == entry.BatchId) { ... } else { throw }`, since we already know that `incomingBatchId >= entry.BatchId` based on L196 as @pranavkm points out.